Element::updateEffectiveTextDirectionIfNeeded recurses children multiple times.

mattwoodrow opened the pull request:
#### e31e3a25e7d493a7d24c88575bcbfbef01bb4c08
<pre>
Element::updateEffectiveTextDirectionIfNeeded recurses children multiple times.
<a href="https://bugs.webkit.org/show_bug.cgi?id=291917">https://bugs.webkit.org/show_bug.cgi?id=291917</a>
&lt;<a href="https://rdar.apple.com/149808238">rdar://149808238</a>&gt;

Reviewed by Said Abou-Hallawa.

updateEffectiveTextDirectionIfNeeded gets called (only) from
insertedIntoAncestor, and does a full descendant walk to compute the direction
if children are present.

insertedIntoAncestor should already be called for all descendants, so we
shouldn&apos;t need to recurse both times and trust that any descendants with auto
direction will be updated in due course.

* Source/WebCore/dom/Element.cpp:
(WebCore::Element::updateEffectiveTextDirectionIfNeeded):

Canonical link: <a href="https://commits.webkit.org/294037@main">https://commits.webkit.org/294037@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/9f7f0b88be9e985eac8f89c4c7652dad183afab0

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/100474 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/20126 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/10425 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/105611 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/51062 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/102515 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/20433 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/28600 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/76506 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/33554 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/103481 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/15658 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/90758 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/56863 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/15475 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/8761 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/50438 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/85390 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/8840 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/107965 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/27592 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/20252 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/85459 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/27955 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/86958 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/84997 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/21660 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/29684 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/7417 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/21557 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/27527 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/32777 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/27338 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/30656 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/28896 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

shallawa commented on Source/WebCore/dom/Element.cpp:
```diff
@@ -2869,7 +2869,7 @@ void Element::updateEffectiveTextDirectionIfNeeded()
         return;
 
-    if (hasAttributeWithoutSynchronization(HTMLNames::dirAttr) || shadowRoot() || firstChild()) {
+    if (hasAttributeWithoutSynchronization(HTMLNames::dirAttr) || shadowRoot()) {
```
Is it guaranteed that when an Element is `insertedIntoAncestor()`, all its descendants will also be re-`insertedIntoAncestor()`?

I tried different scenarios like moving a sub tree from a parent to another parent and removing a sub-tree from the Document and then inserting it back and they seem to work fine.

I could not recall why I added this condition in the first place.

mattwoodrow replied:
It is, the caller iterates all nodes in the subtree.

mattwoodrow committed:
Element::updateEffectiveTextDirectionIfNeeded recurses children multiple times.
https://bugs.webkit.org/show_bug.cgi?id=291917
<rdar://149808238>

Reviewed by Said Abou-Hallawa.

updateEffectiveTextDirectionIfNeeded gets called (only) from
insertedIntoAncestor, and does a full descendant walk to compute the direction
if children are present.

insertedIntoAncestor should already be called for all descendants, so we
shouldn't need to recurse both times and trust that any descendants with auto
direction will be updated in due course.

* Source/WebCore/dom/Element.cpp:
(WebCore::Element::updateEffectiveTextDirectionIfNeeded):

Canonical link: https://commits.webkit.org/294037@main
